File: classic puzzle-medium/sandpile-addition.cpp
//https://www.codingame.com/training/medium/sandpile-addition
//SANDPILE ADDITION
#include <iostream>
#include <cstring>
#include <vector>
#include <algorithm>

using namespace std;
int m[10][10];
//check the grains per pile
bool check_grains(int m[][10],int n){
    for(int i=0;i<n;i++)
        for(int j=0;j<n;j++)
            if(m[i][j]>3) return true;
    return false;
}
int main()
{
    int n;
    cin >> n; cin.ignore();
    memset(m,0,sizeof(n*n));
    string row;
    for (int i = 0; i < n; i++) {
        getline(cin, row);
        for(int j=0;j<row.size();j++)
            m[i][j]=row[j]-'0';
    }
    for (int i = 0; i < n; i++) {
        getline(cin, row);
        for(int j=0;j<row.size();j++){
            m[i][j]+=row[j]-'0';
        }
    }
    
    while(check_grains(m,n)){
        for(int i=0;i<n;i++)
            for(int j=0;j<n;j++){
                if(m[i][j]>3)
                {
                    //add the grains to neighbors
                    int g=m[i][j]/4;
                    m[i][j]%=4;
                    if(i+1<n) m[i+1][j]+=g;
                    if(j+1<n) m[i][j+1]+=g;
                    if(i-1>=0) m[i-1][j]+=g;
                    if(j-1>=0) m[i][j-1]+=g;
                }
            }
        
    }
    for(int i=0;i<n;i++){
        for(int j=0;j<n;j++)
            cout<<m[i][j];
        cout<<endl;
    }
}

File: classic puzzle-medium/sandpile-addition.py
#https://www.codingame.com/training/medium/sandpile-addition
#SANDPILE ADDITION
n=int(input())
m1=[[int(j) for j in input()] for i in range(n)]

for i in range(n):
    r = input()
    for j in range(len(r)):
        m1[i][j]+=int(r[j])

#check the grains per pile
def check_grains(m):
    for i in range(n):
        for j in m[i]:
            if j>=4:
                return False
    return True

while not check_grains(m1):
    for i in range(n):
        for j in range(len(m1)):
            if m1[i][j]>=4:
                g=m1[i][j]//4
                m1[i][j]%=4
                #adding grains to neighbors
                if i+1<n:m1[i+1][j]+=g
                if j+1<n:m1[i][j+1]+=g
                if i-1>=0:m1[i-1][j]+=g
                if j-1>=0:m1[i][j-1]+=g
for x in range(n):
    print(''.join([str(c) for c in m1[x]]))
